### PR TITLE
Add gltf import buffer.

### PR DESCRIPTION
**Pull request by fire**
Part of godotengine/godot-proposals#3273

## How to test?

1. use fox.glb https://github.com/KhronosGroup/glTF-Sample-Models/blob/master/2.0/Fox/glTF-Binary/Fox.glb
2. use fox.gltf
3. use vrm importer with import_buffer patches

### CODE REVIEW

**Comment by fire** [comment]
Discussion this morning with Reduz to unify the Ref<GLTFDocument> api.

* `load_file()`, `load_buffer()` that just generates the internal state, and `save_file` and `save_buffer` functions that save back the files.
* `parse_scene(Node*)` generates the internal state from a Node* scene. And Node* write_scene() writes it out

Expose a `Vector<Ref<GLTFDocumentExtension>>`

**Comment by fire** [comment]
Implementing this design:

* Reading a gltf to a scene

```c++
Ref<GLTFDocument> doc;
//[..]
Error err = doc.load_from_file("path_to_gltf")
//[..]
Node *scene = doc.write_to_scene();
```

* Reading a buffer to a scene

```c++
Ref<GLTFDocument> doc;
//[..]
Error err = doc.load_from_buffer(PackedBytesArray bytes, String p_base_path)
//[..]
Node *scene = doc.write_to_scene();
```

* Writing a gltf to the filesystem

```c++
Ref<GLTFDocument> doc;
// 
Error err = doc.load_from_scene(Node *);
//[..]
Error err = doc.write_to_filesystem("path_to_write")
//[..]
```

* Writing a gltf to a buffer

```c++
Ref<GLTFDocument> doc;
// 
Error err = doc.load_from_scene(Node *);
//[..]
PackedByteArray bytes = doc.write_to_buffer(Error err)
//[..]
```

* doc.get_state() -> GLTFState
* doc.set_state(Ref<GLTFState>) ->state
* add state property

File: editor/import/editor_importer_bake_reset.cpp [hunk 1]
```diff
@@ -75,9 +75,10 @@ void BakeReset::_bake_animation_pose(Node *scene, const String &p_bake_anim) {
 		MeshInstance3D *mesh_3d = scene->cast_to<MeshInstance3D>(node);
 		if (scene->cast_to<Skeleton3D>(node)) {
 			Skeleton3D *skeleton = Object::cast_to<Skeleton3D>(node);
-
-			// Step 2: Bake the RESET animation from the RestBone to the skeleton.
-			_fix_skeleton(skeleton, r_rest_bones);
+			if (skeleton) {
```
Comment (fire) [inline editor/import/editor_importer_bake_reset.cpp:78]: Reminder to move into a separate pr.

**Comment by fire** [comment]
Going to sleep on this and set this ready for review tomorrow.

**Review by lyuma — CHANGES_REQUESTED** [review]

**Review by lyuma — CHANGES_REQUESTED** [review]

**Review by lyuma — APPROVED** [review]

**Comment by fire** [comment]
Crosslinking design:

https://github.com/godotengine/godot-proposals/issues/3273#issuecomment-997403290

```c++
Error append_from_file(String p_path, Ref<GLTFState> r_state, uint32_t p_flags, int32_t p_bake_fps);
Error append_from_buffer(Array p_bytes, String p_base_path, Ref<GLTFState> r_state, uint32_t p_flags, int32_t p_bake_fps);
Error append_from_scene(Node *p_node, Ref<GLTFState> r_state, uint32_t p_flags, int32_t p_bake_fps);
Node *generate_scene(Ref<GLTFState> state, float p_bake_fps);
Array generate_buffer(Ref<GLTFState> state);
Error write_to_filesystem(Ref<GLTFState> state, const String &p_path);
```

File: modules/gltf/gltf_document.cpp [hunk 2]
```diff
@@ -30,6 +30,9 @@
 
 #include "gltf_document.h"
 
+#include "core/error/error_list.h"
```
Comment (fire) [inline modules/gltf/gltf_document.cpp:33]: Will try removing these.

**Review by theraot — CHANGES_REQUESTED** [review]
- For `append_from_buffer` can you keep the order `state`, `flags`, `bake_fps`?
- Please give default values to `flags` and `bake_fps`.
- Appending from file and then saving to file system triggers `ERR_CONTINUE(!mi);` on `_convert_mesh_instances` (otherwise it works correctly). Example code:
```
    	var gltf := GLTFDocument.new()
	var gltf_state := GLTFState.new()
	gltf.append_from_file("res://test.gltf", gltf_state, 0, 0)
	gltf.write_to_filesystem(gltf_state, "res://testx.gltf")
```
- Appending from file and then appending from scene also triggers `ERR_CONTINUE(!mi);` (otherwise it works correctly, checked by `write_to_filesystem`). Example code:
```
	var gltf := GLTFDocument.new()
	var gltf_state := GLTFState.new()
	gltf.append_from_file("res://test.gltf", gltf_state, 0, 0)
	gltf.append_from_scene(self, gltf_state, 0, 0)
```
- Appending from scene and then appending from file results in many `ERR_CONTINUE(state->nodes[child_i]->parent != -1);` on `_parse_nodes` (it works, but it seems to cause some extra duplicates, keeping in mind I'm using an gltf form a similar version of the scene). Example code:
```
	var gltf := GLTFDocument.new()
	var gltf_state := GLTFState.new()
	gltf.append_from_scene(self, gltf_state, 0, 0)
	gltf.append_from_file("res://test.gltf", gltf_state, 0, 0)
```
- Appending buffers twice (the same or different) also triggers many `ERR_CONTINUE(state->nodes[child_i]->parent != -1);` (there seems to be three copies in the result, instead of two). Example code:
```
	var gltf := GLTFDocument.new()
	var gltf_state := GLTFState.new()
	var gltf_state_2 := GLTFState.new()
	gltf.append_from_scene(self, gltf_state, 0, 0)
	var b = gltf.generate_buffer(gltf_state)
	gltf.append_from_buffer(b, "", 0, 0, gltf_state_2)
	gltf.append_from_buffer(b, "", 0, 0, gltf_state_2)
```
- Also appending from file twice (they can be the same file or not) triggers many `ERR_CONTINUE(state->nodes[child_i]->parent != -1);` (otherwise it works correctly). Example code:
```
	var gltf := GLTFDocument.new()
	var gltf_state := GLTFState.new()
	gltf.append_from_file("res://test1.gltf", gltf_state, 0, 0)
	gltf.append_from_file("res://test2.gltf", gltf_state, 0, 0)
```

**Comment by fire** [comment]
Thoughts on a clear gltf state function? Will think about the other changes.

**Comment by theraot** [comment]
> Thoughts on a clear gltf state function?

It is within user expectations. And I say that because the thought crossed my mind while testing.

I'll assume it is a method on `GLTFDocument` given that the others are there.

As per the name, `clear`, `erase`, and `reset`, `wipe` have come to mind, in that order. The name `clear` is the first pick, with the caveat that on `GLTFDocument` it would take a `GLTState` as parameter and clear it… While any other `clear` clears what it is called on and takes no parameters. Do you think another name would work better?

**Comment by fire** [comment]
@theraot  

1. Did some thinking on the clear function. A clear function on a resource is the same as `gltf_document.instantiate()`.
2. I would recommend doing `gltf_document.instantiate()` for `ERR_CONTINUE(state->nodes[child_i]->parent != -1);`. The alternative is some sort of node tree repair..
3. Initialized the flags to 0 because gltf_document is supposed to be possible to be used at runtime and the flags are defined in editor code. Initialized the fps to 30f to match other places.

File: modules/gltf/gltf_document.h [hunk 3]
```diff
@@ -50,6 +51,8 @@
 
 #include <cstdint>
 
+#include "modules/modules_enabled.gen.h"
```
Comment (akien-mga) [inline modules/gltf/gltf_document.h:54]: Check if this is used.

**Review by akien-mga — APPROVED** [review]
Looks fairly good to me, if @godotengine/import folks are happy with it it can be merged.

**Comment by fire** [comment]
🎉

**Pull request merged**